go/vt/vtgate/engine/vindex_lookup.go

```diff
@@ -0,0 +1,245 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+
+	vtgatepb "vitess.io/vitess/go/vt/proto/vtgate"
+
+	"vitess.io/vitess/go/vt/key"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+)
+
+var _ Primitive = (*VindexLookup)(nil)
+
+type VindexLookup struct {
+	Opcode Opcode
+
+	// The vindex to use to do the Map
+	Vindex vindexes.LookupPlanable
+
+	// Keyspace specifies the keyspace to send the query to.
+	Keyspace *vindexes.Keyspace
+
+	Arguments []string
+
+	// Values specifies the vindex values to use for routing.
+	Values []evalengine.Expr
+
+	// We fetch data in order to do the map from this primitive
+	Lookup Primitive
+
+	// This is the side that needs to be routed
+	SendTo *Route
+}
+
+// RouteType implements the Primitive interface
+func (vr *VindexLookup) RouteType() string {
+	return "VindexLookup"
+}
+
+// GetKeyspaceName implements the Primitive interface
+func (vr *VindexLookup) GetKeyspaceName() string {
+	return vr.SendTo.GetKeyspaceName()
+}
+
+// GetTableName implements the Primitive interface
+func (vr *VindexLookup) GetTableName() string {
+	return vr.SendTo.GetTableName()
+}
+
+// GetFields implements the Primitive interface
+func (vr *VindexLookup) GetFields(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return vr.SendTo.GetFields(ctx, vcursor, bindVars)
+}
+
+// NeedsTransaction implements the Primitive interface
+func (vr *VindexLookup) NeedsTransaction() bool {
+	return vr.SendTo.NeedsTransaction()
+}
+
+// TryExecute implements the Primitive interface
+func (vr *VindexLookup) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	ids, err := vr.generateIds(vcursor, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	results, err := vr.lookup(ctx, vcursor, ids)
+	if err != nil {
+		return nil, err
+	}
+
+	dest, err := vr.postLookupPreExecute(ids, results, bindVars)
+	if err != nil {
+		return nil, err
+	}
+
+	return vr.SendTo.executeAfterLookup(ctx, vcursor, bindVars, wantfields, ids, dest)
+}
+
+func (vr *VindexLookup) postLookupPreExecute(ids []sqltypes.Value, results []*sqltypes.Result, bindVars map[string]*querypb.BindVariable) ([]key.Destination, error) {
```

Comment:
nit: this could be named differently - `mapVindexValues` or something would be more descriptive